pkg/authentication/scram/scram.go

```diff
@@ -47,9 +78,250 @@ func EnsureAgentSecret(getUpdateCreator secret.GetUpdateCreator, secretNsName ty
+		return readExistingCredentials(getUpdateCreator, mdbNamespacedName, user.GetUserName())
+	}
+
+	// the password has changed, or we are generating it for the first time
+	sha1Creds, sha256Creds, err := generateScramShaCredentials(user.GetUserName(), password)
+	if err != nil {
+		return scramcredentials.ScramCreds{}, scramcredentials.ScramCreds{}, err
+	}
+
+	// create or update our credentials secret for this user
+	zap.S().Debugf("Generating new credentials and storing in secret/%s", scramCredentialsSecretName(mdbNamespacedName.Name, user.GetUserName()))
+	if err := createScramCredentialsSecret(getUpdateCreator, mdbNamespacedName, user.GetUserName(), sha1Creds, sha256Creds); err != nil {
+		return scramcredentials.ScramCreds{}, scramcredentials.ScramCreds{}, err
+	}
+
+	zap.S().Debugf("Successfully generated SCRAM credentials")
+	return sha1Creds, sha256Creds, nil
+}
+
+// needToGenerateNewCredentials determines if it is required to generate new credentials or not.
+// this will be the case if we are either changing password, or are generating credentials for the first time.
+func needToGenerateNewCredentials(secretGetter secret.Getter, username string, mdbNamespacedName types.NamespacedName, password string) (bool, error) {
+	s, err := secretGetter.GetSecret(types.NamespacedName{Name: scramCredentialsSecretName(mdbNamespacedName.Name, username), Namespace: mdbNamespacedName.Namespace})
+	if err != nil {
+		// haven't generated credentials yet, so we are changing password
+		if errors.IsNotFound(err) {
+			zap.S().Debugf("No existing credentials found, generating new credentials")
+			return true, nil
+		}
+		return false, err
+	}
+
+	existingSha1Salt := s.Data[sha1SaltKey]
+	existingSha256Salt := s.Data[sha256SaltKey]
+
+	// the salts are stored encoded, we need to decode them before we use them for
+	// salt generation
+	decodedSha1Salt, err := base64.StdEncoding.DecodeString(string(existingSha1Salt))
+	if err != nil {
+		return false, err
+	}
+	decodedSha256Salt, err := base64.StdEncoding.DecodeString(string(existingSha256Salt))
+	if err != nil {
+		return false, err
+	}
+
+	// regenerate credentials using the existing salts in order to see if the password has changed.
+	sha1Creds, sha256Creds, err := computeScramShaCredentials(username, password, decodedSha1Salt, decodedSha256Salt)
+	if err != nil {
+		return false, err
+	}
+
+	existingSha1Creds, existingSha256Creds, err := readExistingCredentials(secretGetter, mdbNamespacedName, username)
+	if err != nil {
+		return false, err
+	}
+
+	sha1CredsAreDifferent := sha1Creds != existingSha1Creds
+	sha256CredsAreDifferent := sha256Creds != existingSha256Creds
+
+	return sha1CredsAreDifferent || sha256CredsAreDifferent, nil
+}
+
+// generateSalts generates 2 different salts. The first is for the sha1 algorithm
+// the second is for sha256
+func generateSalts() ([]byte, []byte, error) {
```

Comment:
do these belong to the `generate` package?

Reply:
yes I think you're right!